### PR TITLE
Index scheduler cancel tasks

### PR DESCRIPTION
**Pull request by Kerollmops**
This PR has been rebased on the [*index-scheduler-delete-persisted-file](https://github.com/meilisearch/meilisearch/tree/index-scheduler-delete-persisted-file) branch and should not be merged for now.

What does this PR do:
 - It introduces the tasks cancelation operation that can be triggered by using the `POST /tasks/cancel?uid=1,2,3&status=enqueued&type=documentAdditionOrUpdate&indexUid=movies` route.
 - It can cancel processing or enqueued tasks and will change a `must_stop` boolean internally to inform a possible indexation to stop as quickly as possible.
- It adds a new `canceledBy` field to the `TaskView` and the `DumpTaskView`.
- It also supports dumping tasks with a canceled status and the `canceledBy` field.

We need to add more tests, but overall it looks good.

### CODE REVIEW

File: index-scheduler/src/lib.rs [hunk 1]
```diff
@@ -381,11 +424,21 @@ impl IndexScheduler {
         match wtxn.commit() {
             Ok(()) => (),
             _e @ Err(_) => {
-                todo!("remove the data associated with the task");
+                self.delete_persisted_task_data(&task)?;
                 // _e?;
```
Comment (irevoire) [inline index-scheduler/src/lib.rs:428]: We should throw the error after that

File: index-scheduler/src/lib.rs [hunk 2]
```diff
@@ -606,6 +676,25 @@ impl IndexScheduler {
...
+            | KindWithContent::IndexDeletion { .. }
+            | KindWithContent::IndexCreation { .. }
+            | KindWithContent::IndexUpdate { .. }
+            | KindWithContent::IndexSwap { .. }
+            | KindWithContent::TaskCancelation { .. }
+            | KindWithContent::TaskDeletion { .. }
+            | KindWithContent::DumpExport { .. }
+            | KindWithContent::Snapshot => Ok(()),
+        }
+    }
```
Comment (irevoire) [inline index-scheduler/src/lib.rs:697]:
```suggestion
    if let Some(content_file) = task.content_uuid() {
        self.delete_update_file(*content_file)?;
    }
    Ok(())
```

We have a function for that

File: index-scheduler/src/utils.rs [hunk 3]
```diff
@@ -2,24 +2,18 @@
...
-            Status::Processing,
-            Status::Succeeded,
-            Status::Failed,
-        ] {
-            all_tasks |= self.get_status(&rtxn, status)?;
-        }
-        Ok(all_tasks)
+        enum_iterator::all()
+            .map(|s| self.get_status(&rtxn, s))
+            .r#union()
```
Comment (irevoire) [inline index-scheduler/src/utils.rs:14]:
```suggestion
            .union()
```

We don't need anything specific to handle the handle the `union` word as a method?

**Review by irevoire — CHANGES_REQUESTED** [review]

File: index-scheduler/src/lib.rs [hunk 4]
```diff
@@ -117,6 +119,40 @@ impl Query {
...
+}
+
+impl ProcessingTasks {
+    /// Stores the currently processing tasks, the date time at which it started
+    /// and resets the _must stop_ flag.
+    fn start_processing_at(&mut self, started_at: OffsetDateTime, processing: RoaringBitmap) {
+        self.started_at = started_at;
+        self.processing = processing;
+        self.must_stop.store(false, Relaxed);
+    }
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply index-scheduler/src/lib.rs]: We can't do so as the date must be known by the `tick` method for when it finishes (step 2.).

File: index-scheduler/src/utils.rs (hunk 3 above)
Comment (Kerollmops) [reply index-scheduler/src/utils.rs:14]: But it breaks my editor syntax highlighting 😢 Ok... I'll remove that...

**Pull request merged**